### PR TITLE
Verify supported Elasticsearch distribution during license reconciliation

### PR DESCRIPTION
**Pull request by thbkrkr**
This commit makes ECK more strict against ES license reconciliation errors to verify that it is managing a supported Elasticsearch distribution.

This produces events:
>    Warning  Unexpected               7s (x14 over 54s)    elasticsearch-controller              Unsupported Elasticsearch distribution: unable to verify Elasticsearch license: 400 Bad Request: Invalid index name [_license], must not start with '_'.

And logs:
```
2021-10-04T16:17:58.683+0200    ERROR   manager.eck-operator.controller.elasticsearch-controller        Reconciler error        {
    "service.version": "1.9.0-SNAPSHOT+ec9237d8",
    "name": "os",
    "namespace": "default",
    "error": "unsupported elasticsearch distribution: unable to verify Elasticsearch license: 400 Bad Request: unknown",
    "errorCauses": [
        {
            "error": "unsupported elasticsearch distribution: unable to verify Elasticsearch license: 400 Bad Request: unknown",
            "errorVerbose": "400 Bad Request: unknown
            unable to verify Elasticsearch license
            github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/license.checkElasticsearchLicense
            github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/license.Reconcile
            github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver.(*defaultDriver).Reconcile
            github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch.(*ReconcileElasticsearch).internalReconcile
            github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch.(*ReconcileElasticsearch).Reconcile
            sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
            sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
            sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
            runtime.goexit
            unsupported elasticsearch distribution
            github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver.(*defaultDriver).Reconcile
            github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch.(*ReconcileElasticsearch).internalReconcile
            github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch.(*ReconcileElasticsearch).Reconcile
            sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
            sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
            sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
            runtime.goexit
        }
    ]
}
```

### CODE REVIEW

File: pkg/controller/elasticsearch/license/reconcile.go [hunk 1]
```diff
@@ -18,7 +18,25 @@ func Reconcile(
...
+	}
+
 	clusterName := k8s.ExtractNamespacedName(&esCluster)
-	return applyLinkedLicense(ctx, c, clusterName, clusterClient)
+	return true, applyLinkedLicense(ctx, c, clusterName, clusterClient, currentLicense)
+}
+
+// checkElasticsearchLicense checks that Elasticsearch is licensed, which ensures that the operator is communicating
+// with a supported version of Elasticsearch
+func checkElasticsearchLicense(ctx context.Context, clusterClient esclient.LicenseClient) (esclient.License, bool, error) {
```
[the remark this answers is not in the document]
Comment (thbkrkr) [reply pkg/controller/elasticsearch/license/reconcile.go:33]: This makes unit testing easier.

File: pkg/controller/elasticsearch/license/reconcile.go [hunk 2]
```diff
@@ -18,7 +18,25 @@ func Reconcile(
...
+	currentLicense, err := clusterClient.GetLicense(ctx)
+	if err != nil {
+		// 4xx is not supported, except 404 which may happen if the master node is generating a new cluster state
+		unsupportedElasticsearch := esclient.Is4xx(err) && !esclient.IsNotFound(err)
```
[the remark this answers is not in the document]
Comment (thbkrkr) [reply pkg/controller/elasticsearch/license/reconcile.go:37]: Yes, let's make sure we don't lock the cluster down for these errors.

**Comment by thbkrkr** [comment]
Jenkins test this please

File: pkg/controller/elasticsearch/license/reconcile.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pebrc) [reply pkg/controller/elasticsearch/license/reconcile.go:34]: Makes sense 👍

File: pkg/controller/elasticsearch/driver/driver.go [hunk 3]
```diff
@@ -194,16 +196,20 @@ func (d *defaultDriver) Reconcile(ctx context.Context) *reconciler.Results {
...
+	if esReachable { //nolint:nestif
+		// reconcile the Elasticsearch license
+		supportedDistribution, err := license.Reconcile(ctx, d.Client, d.ES, esClient)
+		if err != nil {
+			if !supportedDistribution {
+				msg := "Unsupported Elasticsearch distribution"
+				d.ReconcileState.AddEvent(corev1.EventTypeWarning, events.EventReasonUnexpected, fmt.Sprintf("%s: %s", msg, err.Error()))
+				// unsupported distribution, let's update the phase to "invalid" and stop the reconciliation
+				d.ReconcileState.UpdateElasticsearchStatusPhase(esv1.ElasticsearchResourceInvalid)
+				return results.WithError(errors.Wrap(err, strings.ToLower(msg)))
```
Comment (pebrc) [inline pkg/controller/elasticsearch/driver/driver.go:208]: This also lower cases "Elasticsearch"

File: pkg/controller/elasticsearch/driver/driver.go [hunk 4]
```diff
@@ -194,16 +196,20 @@ func (d *defaultDriver) Reconcile(ctx context.Context) *reconciler.Results {
...
-				events.EventReasonUnexpected,
-				fmt.Sprintf("Could not update cluster license: %s", err.Error()),
-			)
-			log.Info("Could not update cluster license", "err", err, "namespace", d.ES.Namespace, "es_name", d.ES.Name)
-			// don't error out the entire reconciliation, move on with next steps and retry later
+	if esReachable { //nolint:nestif
+		// reconcile the Elasticsearch license
+		supportedDistribution, err := license.Reconcile(ctx, d.Client, d.ES, esClient)
+		if err != nil {
+			if !supportedDistribution {
```
Comment (pebrc) [inline pkg/controller/elasticsearch/driver/driver.go:203]: I think we can remove one level of nesting here:
```golang
supportedDistribution, err := license.Reconcile(ctx, d.Client, d.ES, esClient)
if !supportedDistribution {} // implies err != nil but we add that to condition to make it more explicit
if err != nil {}
```

**Review by pebrc — COMMENTED** [review]
Found a few nits but otherwise LGTM 👍

File: pkg/controller/elasticsearch/driver/driver.go (hunk 4 above)
Comment (thbkrkr) [reply pkg/controller/elasticsearch/driver/driver.go:203]: Indeed, it's better but I'd stick to the explicit way.
```
supportedDistribution, err := license.Reconcile(ctx, d.Client, d.ES, esClient)
if err != nil && !supportedDistribution {}
if err != nil {}
```

**Pull request merged**